src/Components/Server/src/Circuits/RemoteRenderer.cs

```diff
@@ -207,22 +212,33 @@ public void OnRenderCompleted(long incomingBatchId, string errorMessageOrNull)
             // line (i.e., matching the order in which we received batch completion messages) based on the fact that SignalR
             // synchronizes calls to hub methods. That is, it won't issue more than one call to this method from the same hub
             // at the same time on different threads.
-            if (!PendingRenderBatches.TryDequeue(out var entry) || entry.BatchId != incomingBatchId)
+            if (!PendingRenderBatches.TryDequeue(out var entry))
             {
                 HandleException(
                     new InvalidOperationException($"Received a notification for a rendered batch when not expecting it. Batch id '{incomingBatchId}'."));
             }
             else
             {
-                if (errorMessageOrNull == null)
+                entry.Data.Dispose();
```

Comment:
It seems strange that we're doing there here, and not (say) just before L242. If `(entry.BatchId != incomingBatchId)` why would we still dispose `entry.Data` - isn't that disposing a random other array?

Reply:
@javiercn is looking in to more of this, but as of writing this, we effectively terminate the connection once we get an out of order ACK (since we throw once that happens). We couldn't defer it since the renderer dispose doesn't have a way to get to the dequeued instance to Dispose at that point.

Reply:
I'm updating this exact code as we speak, we can do it differently, with try peak above. I'll send a PR out in like 30 min (adding the tests) so you see